hydra-node/test/Hydra/Model/MockChain.hs

```diff
@@ -0,0 +1,198 @@
+import qualified Data.Sequence.Strict as StrictSeq
+import Hydra.BehaviorSpec (
+  ConnectToChain (..),
+ )
+import Hydra.Chain (Chain (..))
+import Hydra.Chain.Direct.Fixture (testNetworkId)
+import Hydra.Chain.Direct.Handlers (ChainSyncHandler, DirectChainLog, SubmitTx, chainSyncHandler, mkChain, onRollForward)
+import Hydra.Chain.Direct.ScriptRegistry (ScriptRegistry (..))
+import Hydra.Chain.Direct.State (ChainContext (..), ChainStateAt (..))
+import qualified Hydra.Chain.Direct.State as S
+import Hydra.Chain.Direct.TimeHandle (TimeHandle)
+import qualified Hydra.Chain.Direct.Util as Util
+import Hydra.Chain.Direct.Wallet (TinyWallet (..))
+import Hydra.Crypto (HydraKey)
+import Hydra.HeadLogic (
+  Environment (party),
+  Event (NetworkEvent),
+  HeadState (..),
+  defaultTTL,
+ )
+import Hydra.Ledger.Cardano (genTxIn)
+import Hydra.Logging (Tracer)
+import Hydra.Model.Payment (CardanoSigningKey (..))
+import Hydra.Network (Network (..))
+import Hydra.Network.Message (Message)
+import Hydra.Node (
+  HydraNode (..),
+  chainCallback,
+  createNodeState,
+  putEvent,
+ )
+import Hydra.Party (Party (..), deriveParty)
+import Ouroboros.Consensus.Cardano.Block (HardForkBlock (..))
+import qualified Ouroboros.Consensus.Protocol.Praos.Header as Praos
+import Ouroboros.Consensus.Shelley.Ledger (mkShelleyBlock)
+import Test.Consensus.Cardano.Generators ()
+
+-- | Provide the logic to connect a list of `MockHydraNode` through a dummy chain.
+mockChainAndNetwork ::
+  forall m.
+  ( MonadSTM m
+  , MonadTimer m
+  , MonadThrow m
+  , MonadAsync m
+  , MonadThrow (STM m)
+  , MonadLabelledSTM m
+  ) =>
+  Tracer m DirectChainLog ->
+  [(SigningKey HydraKey, CardanoSigningKey)] ->
+  [Party] ->
+  TVar m [MockHydraNode m] ->
+  m (ConnectToChain Tx m, Async m ())
+mockChainAndNetwork tr seedKeys _parties nodes = do
+  queue <- newTQueueIO
+  labelTQueueIO queue "chain-queue"
+  tickThread <- async (labelThisThread "chain" >> simulateTicks queue)
+  let chainComponent = \node -> do
+        let ownParty = party (env node)
+        let (vkey, vkeys) = findOwnCardanoKey ownParty seedKeys
+        let ctx =
+              S.ChainContext
+                { networkId = testNetworkId
+                , peerVerificationKeys = vkeys
+                , ownVerificationKey = vkey
+                , ownParty
+                , scriptRegistry =
+                    -- TODO: we probably want different _scripts_ as initial and commit one
+                    let txIn = mkMockTxIn vkey 0
+                        txOut =
+                          TxOut
+                            (mkVkAddress testNetworkId vkey)
+                            (lovelaceToValue 10_000_000)
+                            TxOutDatumNone
+                            ReferenceScriptNone
+                     in ScriptRegistry
+                          { initialReference = (txIn, txOut)
+                          , commitReference = (txIn, txOut)
+                          }
+                }
```

Comment:
Should: consider using generators here.

You could define a `HydraContext` based on the `seedKeys` and then use [`deriveChainContexts`](https://github.com/input-output-hk/hydra/blob/e3c73d275e6aec1f10711cf385a97fe800759aaf/hydra-node/src/Hydra/Chain/Direct/State.hs#L763-L764) to get the individual `ChainContext` values.


Or at least use `genScriptRegistry` :)

Reply:
Then you can remove the TODO here and the "NOTE: Ugly," below

Reply:
That's a good idea, but to implement it properly would require some significant refactoring which I am willing to address in another PR: Is that ok?